### ISSUE TITLE
Some questions on the N-step ReplayBuffer

### ISSUE THREAD

**Issue by ty2000** [Curt-Park/rainbow-is-all-you-need#10]
Maybe some dumb questions about the N-step ReplayBuffer

1. In update_model()

```
        if self.use_n_step:
            samples = self.memory_n.sample_batch_from_idxs(indices)
            gamma = self.gamma ** self.n_step
            n_loss = self._compute_dqn_loss(samples, gamma)
```

Here the assumption is in the samples, for each transition, the next_obs will be n_step away from the obs so self.gamma ** self.n_step makes sense. However, the way _get_n_step_info() is implemented, there is no guarantee that assumption is true, for example when there is a terminal state right after obs, in that case, for that transition, the next_obs will be just 1-step away from the obs, right?

2. In the same code snippet above
`samples = self.memory_n.sample_batch_from_idxs(indices)`
Here self.memory_n is samped with same indices sampled with self.memory. However, there seems two questions here
2.1 For the same index, the two transitions sampled from self.memory and self.memory_n respectively do not have the same obs (they have same next_obs if there is no terminal state issue, as described above)
2.2 If there is a terminal state after obs but before the n-step, then the transitions sampled from self.memory and self.memory_n have neither same obs nor same next_obs

For example, the following was from a trace (the samples from N-step is renamed to samples2 for clarity)

(Pdb) l
138  	        # prevent high-variance.
139  	        if self.use_n_step:
140  	            samples2 = self.memory_n.sample_batch_from_idxs(indices)
141  	            if (samples2['next_obs'][0][0] != samples['next_obs'][0][0]):
142  	              pdb.set_trace()
143  ->	            gamma = self.gamma ** self.n_step
144  	            n_loss = self._compute_dqn_loss(samples2, gamma)
145  	            loss += n_loss
146  	
147  	        self.optimizer.zero_grad()
148  	        loss.backward()
(Pdb) p samples
{'obs': array([[-0.03751984,  0.00060021,  0.00848503,  0.02023765]],
      dtype=float32), 'next_obs': array([[-0.03750784, -0.1946424 ,  0.00888978,  0.31558558]],
      dtype=float32), 'acts': array([0.], dtype=float32), 'rews': array([1.], dtype=float32), 'done': array([0.], dtype=float32), 'indices': array([17])}
(Pdb)  samples2
{'obs': array([[-0.04211658, -0.7778139 ,  0.17969099,  1.5349392 ]],
      dtype=float32), 'next_obs': array([[-0.05767286, -0.974587  ,  0.21038976,  1.8778918 ]],
      dtype=float32), 'acts': array([0.], dtype=float32), 'rews': array([1.], dtype=float32), 'done': array([1.], dtype=float32)}

So my question is, if the samples from 1-step and N-step are not really synchronized, what is the benefit of trying? I guess even if we simply use self.memory_n.sample_batch the result is probably going to be fine since here the 1-step loss is just used to lower the variance, and that does not necessary require the two samples to have the same indices?

Thanks!

**Comment by MrSyee** [comment]
Hi, @ty2000 !

Thank you for your feedback. I answer your questions.

1. When the next_obs is less than n-step, for example just 2 step, done is True by _get_n_step_info(). Then reward of n-step buffer is $R_1 + \gamma R_2$, and gamma * next_q_value * mask is 0 by mask (1 - done) in _compute_dqn_loss(). target is just reward, that is $R_1 + \gamma R_2$. For this reason, it doesn't matter if it's gamma ** self.n-step.

2. It is right that you said in 2.1. Current obs must be the same in the 1-step buffer and n-step buffer when the same index. We'll consider how this part does modify. However, it is just OK that 1-step buffer and n-step buffer are only synchronized at the current obs. Because we want that it is updated about the Q value of current obs.

Thanks.

**Comment by ty2000** [comment]
Thanks for the clarification. You are right on question 1 and 2.1, I missed the mask part. 

As for 2.1, I also think ideally obs in 1-step and n-step should be the same, however the implementation right now actually has same next_obs, instead of obs. maybe when sampling from n-step buffer using the index, the index needs to be adjusted backward to retrieve the correct obs

**Comment by Curt-Park** [comment]
@MrSyee Please review the pull-request.

**Comment by MrSyee** [comment]
@Curt-Park I'd checked this PR and merged. Thank you for your prompt action!

@ty2000 You're right. We fixed this problem. Can you check this code again? Thanks.
